packages/grpc-js/src/server-call.ts

```diff
@@ -553,105 +553,103 @@ export class Http2ServerCallStream<
-    const body: Buffer[] = [];
-    const limit = this.maxReceiveMessageSize;
+      function onData(chunk: Buffer) {
+        receivedLength += chunk.byteLength;
 
-    stream.on('data', onData);
-    stream.on('end', onEnd);
-    stream.on('error', onEnd);
+        if (limit !== -1 && receivedLength > limit) {
+          stream.removeListener('data', onData);
+          stream.removeListener('end', onEnd);
+          stream.removeListener('error', onEnd);
+
+          reject({
+            code: Status.RESOURCE_EXHAUSTED,
+            details: `Received message larger than max (${receivedLength} vs. ${limit})`,
+          });
+          return;
+        }
 
-    function onData(chunk: Buffer) {
-      receivedLength += chunk.byteLength;
+        body.push(chunk);
+      }
 
-      if (limit !== -1 && receivedLength > limit) {
+      function onEnd(err?: Error) {
         stream.removeListener('data', onData);
         stream.removeListener('end', onEnd);
         stream.removeListener('error', onEnd);
-        next({
-          code: Status.RESOURCE_EXHAUSTED,
-          details: `Received message larger than max (${receivedLength} vs. ${limit})`,
-        });
-        return;
-      }
 
-      body.push(chunk);
-    }
-
-    function onEnd(err?: Error) {
-      stream.removeListener('data', onData);
-      stream.removeListener('end', onEnd);
-      stream.removeListener('error', onEnd);
+        if (err !== undefined) {
+          reject({ code: Status.INTERNAL, details: err.message });
+          return;
+        }
 
-      if (err !== undefined) {
-        next({ code: Status.INTERNAL, details: err.message });
-        return;
-      }
+        if (receivedLength === 0) {
+          reject({
+            code: Status.INTERNAL,
+            details: 'received empty unary message',
+          });
+          return;
+        }
 
-      if (receivedLength === 0) {
-        next({
-          code: Status.INTERNAL,
-          details: 'received empty unary message',
-        });
-        return;
-      }
+        call.emit('receiveMessage');
 
-      call.emit('receiveMessage');
+        const requestBytes = Buffer.concat(body, receivedLength);
+        const compressed = requestBytes.readUInt8(0) === 1;
+        const compressedMessageEncoding = compressed ? encoding : 'identity';
+        const decompressedMessage = call.getDecompressedMessage(
+          requestBytes,
+          compressedMessageEncoding
+        );
 
-      const requestBytes = Buffer.concat(body, receivedLength);
-      const compressed = requestBytes.readUInt8(0) === 1;
-      const compressedMessageEncoding = compressed ? encoding : 'identity';
-      const decompressedMessage = call.getDecompressedMessage(
-        requestBytes,
-        compressedMessageEncoding
-      );
+        if (Buffer.isBuffer(decompressedMessage)) {
+          call
+            .deserializeMessageWithInternalError(decompressedMessage)
+            .then(resolve)
+            .catch(reject);
```

Comment:
I'm pretty sure this can just be `resolve(call.deserializeMessageWithInternalError(decompressedMessage))`. The output of that function is now a promise-like object, which can be passed directly to `resolve`.

Reply:
I actually preferred the readability of explicitly passing resolve and reject, to actually show why we might want a Promise here, but I can also understand why you would want to keep this as simple as possible.
I'll push a change for that as well.

Reply:
I'm not really concerned, it's still very readable. I just personally prefer writing explicit code, but as I said I also understand why keeping it simplistic has its benefits, and I definitely want to adhere to the code style this repository aims to maintain.